src/platform_impl/web/custom_cursor.rs

```diff
@@ -0,0 +1,56 @@
+use std::sync::Arc;
+
+use wasm_bindgen::{Clamped, JsCast};
+use web_sys::{CanvasRenderingContext2d, HtmlCanvasElement};
+
+use crate::custom_cursor::BadCursor;
+
+#[derive(Debug, Clone)]
+pub struct WebCustomCursor {
+    pub(crate) inner: Arc<str>,
+}
+
+impl WebCustomCursor {
+    pub fn from_rgba(
+        rgba: Vec<u8>,
+        width: u32,
+        height: u32,
+        hotspot_x: u32,
+        hotspot_y: u32,
+    ) -> Result<Self, BadCursor> {
+        #[allow(clippy::disallowed_methods)]
+        let cursor_icon_canvas = web_sys::window()
```

Comment:
These functions are disallowed for a reason (the reason is also stated in the warning).
Please pass these through appropriately.

Reply:
I would really love to have an unified constructor for all platforms. Passing in extra stuffy only for web would be really annoying. If we need to do that, then the interface seems really inconvinient for web in addition to linux.

Reply:
I don't think this necessarily changes the public API, it just complicates the API internally.
You will have to split construction in two: one is the public part, which doesn't do much apart from verifying sizes and such. The other part actually builds the cursor and data URL when `CustomCursor` is eventually passed to `Window::set_custom_cursor()`.

The way it's done right now is already an issue exactly because of what the warning was added to prevent: any user calling `CustomCursor` outside the main thread will experience a panic completely circumventing the wire implementation we have in place exactly for these kind of reasons.

Reply:
So what you are basically saying is that canvas images can only be created in the window struct. Like here in the other branch. https://github.com/eero-lehtinen/winit/blob/55e2cbb4d569d948f964e66b71e52ecf6ef56725/src/platform_impl/web/window.rs#L225. The reason I originally made it that way was that it seemed too complicated to to let users create cursors, which seems to be confirmed now.

Reply:
`HTMLCanvasElement` can only be created in the Web Window, that is the reason why we disallow creating a Web `Window` to prevent issues like that. It can be done without the Winit `Window` of course, but you will have to implement thread detection logic.

I think the way forward for now is indeed what you have done in this other branch.

As a follow-up PR we can do the following:
- Create some sort of shared `Global` struct to de-duplicate thread detection logic. Use that to actually move the `Canvas` instantiation to `CustomCursor::from_rgba()`.
- Use `OffscreenCanvas` when available, which we need to do anyway when we are in a thread.
- (Bonus) Add a platform-specific extension to create custom icons by URL.

Reply:
Code moved to src/platform_impl/web/custom_cursor.rs. Now is called safely from the Window struct and uses document saved in it.

I also added the possibility to create url icons.